preprocess.py

```diff
@@ -1,255 +1,286 @@
+        trained_features = list(chain.from_iterable(self.trained_features.values()))
+        adjusted_features_df = self.adjust_features(dummies_df, trained_features)
+        return adjusted_features_df
+
+    # averaged over the 'cabin_indexes' feature to have a single numeric value
+    # missing cells were imputed with a constant value
+    # input:    df: pd.DataFrame - target df
+    # output:   avg_cabin_index_df: pd.DataFrame - transformed df
+    def cabin_indexes_impute(self, df: pd.DataFrame) -> pd.DataFrame:
+        def avg_cabin_index(item):
+            if type(item) == list:
+                if len(item) == 0:
+                    return self.missing_indexes_value
+            return np.mean(item)
+
+        avg_cabin_index_df = df.copy()
+        avg_cabin_index_df['cabin_indexes'] = avg_cabin_index_df['cabin_indexes'].fillna(self.missing_indexes_value)
+        avg_cabin_index_df['cabin_indexes'] = avg_cabin_index_df['cabin_indexes'].apply(avg_cabin_index)
+
+        return avg_cabin_index_df
+
+    # encoded the missingness of 'cabin_type' using an additional label - 'Missing'
+    def cabin_type_impute(self, df: pd.DataFrame) -> pd.DataFrame:
+        missingness_label = 'Missing'
+        cabin_type_missing_label_df = df.copy()
+        missing_idx = cabin_type_missing_label_df['cabin_type'][cabin_type_missing_label_df['cabin_type'].isna()].index
+
+        cabin_type_missing_label_df.loc[missing_idx, 'cabin_type'] = missingness_label
+
+        return cabin_type_missing_label_df
+
+    # adjusts the columns of the df for the DataFrame of val/test sets
+    # This is needed as we're not guaranteed to get the exact same subset
+    # of features when using pd.get_dummies on the val/test set, since not all
+    # values that appear in one df necessarily appear in the other df.
+    # input:    df: pd.DataFrame - the df to be adjusted
+    #           trained_features: dict - subset of features which the train set had after pd.get_dummies
+    # output:   X_missing: pd.DataFrame - adjusted df
+    def adjust_features(self, df: pd.DataFrame, trained_features: dict) -> pd.DataFrame:
+        X_missing = df.copy()
+        # dropping features which the imputation model wasn't trained on
+        features_to_drop = set(X_missing.columns) - set(trained_features)
+        features_to_drop = [item for item in features_to_drop]
+        X_missing = X_missing.drop(features_to_drop, axis=1)
+
+        # adding features which the imputation model was trained on and are missing in the test set
+        features_to_add = set(trained_features) - set(X_missing.columns)
+        features_to_add = [item for item in features_to_add]
+        X_missing[features_to_add] = 0
+
+        # reordering features to match the input order into the imputation model
+        X_missing = X_missing[trained_features]
+
+        return X_missing
+
+
+# TODO: choose a method for feature selection and implement it

```

Comment:
For the future.. If there is a feature that you didn't implement yet, it shouldn't be in your PR...
When you split this classes into different files you could just not include this file in your commit...